r2-shared/src/main/java/org/readium/r2/shared/fetcher/Resource.kt

```diff
@@ -169,6 +171,76 @@ abstract class ProxyResource(protected val resource: Resource) : Resource {
     override suspend fun close() = resource.close()
 }
 
+class CachingResource(protected val resource: Resource) : Resource {
+
+    private lateinit var _link: Link
+    private lateinit var _length: ResourceTry<Long>
+    private lateinit var _bytes: ResourceTry<ByteArray>
+
+    override suspend fun link(): Link {
+        if (!::_link.isInitialized) {
+            _link = resource.link()
+        }
+        return _link
+    }
+
+    override suspend fun length(): ResourceTry<Long> {
+        if (!::_length.isInitialized) {
+            _length = resource.length()
+        }
+        return _length
+    }
+
+    override suspend fun read(range: LongRange?): ResourceTry<ByteArray> {
+        if (!::_bytes.isInitialized) {
+            _bytes = resource.read()
+        }
+
+        if (range == null)
+            return _bytes
+
+        @Suppress("NAME_SHADOWING")
+        val range = range
+            .coerceToPositiveIncreasing()
+            .requireLengthFitInt()
+        return _bytes.map { it.sliceArray(range.map(Long::toInt)) }
+    }
+
+    override suspend fun close() = resource.close()
+}
```

Comment:
Inspired from `InputStream.buffered()` which returns a `BufferedInputStream`.

Could be used to cache an expensive resource (e.g. `TransformingResource`) in a nicer way.

```suggestion
}

/**
 * Creates a cached resource wrapping this resource.
 */
fun Resource.cached(): Resource =
    if (this is CachingResource) this
    else CachingResource(this)
```